### PR TITLE
Update Discovery Gateway and make vtgatev2_test pass.

### PR DESCRIPTION
**Pull request by guoliang100**
@alainjobart @sougou 

Also fix many pylint errors.

### CODE REVIEW

File: go/cmd/vtcombo/main.go [hunk 1]
```diff
@@ -36,6 +40,7 @@ var (
 	shardingColumnName = flag.String("sharding_column_name", "keyspace_id", "Specifies the column to use for sharding operations")
 	shardingColumnType = flag.String("sharding_column_type", "", "Specifies the type of the column to use for sharding operations")
 	vschema            = flag.String("vschema", "", "vschema file")
+	tabletTypesToWait  = flag.String("tablet_types_to_wait", "", "wait till connected for specified tablet types during Gateway initialization")
```
Comment (alainjobart) [inline go/cmd/vtcombo/main.go]: For vtcombo, I would just hard-code master+replica+rdonly in the vtgate.Init call. We create one of each for each shard, and we want to wait for them all.

File: go/vt/vtgate/discoverygateway.go [hunk 2]
```diff
@@ -30,38 +33,120 @@ var (
...
+			allShards := make(map[string]bool)
+			for _, ksPartition := range ks.Partitions {
+				for _, shard := range ksPartition.ShardReferences {
+					allShards[shard.Name] = true
+				}
+			}
+			// check connections
+			for shardName := range allShards {
+				wg.Add(1)
+				go func(shard string) {
```
Comment (alainjobart) [inline go/vt/vtgate/discoverygateway.go]: this go func is not very useful and can be removed, as we don;t do any processing in there, just loop over entries in an array. The inner go func() will do the background work.

**Comment by alainjobart** [comment]
I am not sure the discovery parameter deserves its own protocol_flavors, as it's not really a different protocol. To me, it's more like a MariaDB vs MySQL parameter, or ZK vs etcd. Do you mind looking at making it that way, with a new command line flag in utils.py?

**Comment by alainjobart** [comment]
Also, in go/vt/tabletmanager/binlog.go, around line 280, there is code to wait for an initial endpoint with a timeout. Could that code also be using the same thing you are using in this CL to wait for endpoints?

File: go/cmd/vtcombo/main.go (hunk 1 above)
Comment (guoliang100) [inline go/cmd/vtcombo/main.go]: Updated.

File: go/vt/vtgate/discoverygateway.go (hunk 2 above)
Comment (guoliang100) [inline go/vt/vtgate/discoverygateway.go]: Updated.

**Comment by guoliang100** [comment]
Changed the gateway flavor to option. Added couple of flavor classes so we can have gateway specific flags and setup work.

**Comment by guoliang100** [comment]
The waitForEndPoints() needs to wait for at least one endpoint for every tablet type.
Blocking on StateUpdate only guarantees one endpoint for the shard (but any tablet type).

**Comment by alainjobart** [comment]
Well in my use case I can pass in [replica] for the tablet types, and then it will wait for at least one tablet of type replica, which is what I want... But you can submit this change, and then I can make the change on my side.

LGTM.

**Pull request merged**